### ISSUE TITLE
Re-evaluate use of useCallback and related performance practices

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#2351]
## Feature Description

Prerequisite reading: 
* [`useCallback`](https://reactjs.org/docs/hooks-reference.html#usecallback)
* [React, Inline Functions, and Performance](https://reacttraining.com/blog/react-inline-functions-and-performance/#purecomponent-and-shouldcomponentupdate)

This issue is primarily about our use of `useCallback` but the considerations apply to a few other places as well, such as our use of `Function.bind` in legacy components. If you haven't already, please read the above prerequisites before continuing.

In the Site Kit codebase, we primarily use `useCallback` to wrap functions that are passed as props to a component. Early on, we made this our default approach which probably made more sense at the time as we had more legacy components around.

The TL;DR here is that this wrapping is unnecessary more often than not and may result in sub-optimal performance because of all the extra work done by using it. It may also lead to bugs due to complications around callback dependencies and including them all in the list of dependencies for the hook. While many other hooks accept a list of dependencies, it's particularly important for this one because it affects what's in scope for the wrapped function.

From the docs for `useCallback`:
> This is useful when passing callbacks to optimized child components that rely on reference equality to prevent unnecessary renders (e.g. `shouldComponentUpdate`).

This is the primary reason to use such a technique – where referential equality is important between renders.

i.e. 
```jsx
<SomeComponent onUpdate={ () => doSomething() } />
```
In this example, the function passed to `onUpdate` will be different on every render because it's defined inline. For most components, this is fine. This only affects class components that implement custom logic for `shouldComponentUpdate`, components extending [`PureComponent`](https://reactjs.org/docs/react-api.html#reactpurecomponent), or that are wrapped with [`React.memo`](https://reactjs.org/docs/react-api.html#reactmemo).

Since we primarily use functional components now this is no longer a concern, at least for our 1st party components.

There _may_ still be a case for using these when passing to 3rd party components, but the bottom line is that without measuring, we don't know if these performance optimizations are actually more performant or not.

The same goes for [function binding in legacy component constructors](https://github.com/google/site-kit-wp/blob/e1bd73a71e81ce09aab24e29691da6d8ddb2ab07/assets/js/components/data-block.js#L42-L43) (which actually seems to be worse performance-wise), although these will naturally go away as we refactor them.

Let's re-evaluate our use of `useCallback` and see if and where it makes sense to keep. Otherwise, we may be making things more complicated and potentially slower than necessary.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation Brief

* 

### Test Coverage

* 

### Visual Regression Changes

* 

## QA Brief

* 

## Changelog entry

*

**Comment by eugene-manuilov** [comment]
@aaemnnosttv thanks for creating this ticket, but I think that we need to clearly identify "performance" meaning and a metric behind it before we start suggesting improvements. Which performance metric do you mean (TTI, FMP, FCP, etc) and how much do you think it is affected by `useCallback` and functions binding usage? I just want to make sure that we aren't going to spend a few dozen hours on reworking callback functions to make our scripts more performant on 0.00001%.

**Comment by tofumatt** [comment]
Yeah, that's a good call from @eugene-manuilov. It might make sense to, going forward, not bother adding `useCallback` where we don't need it—but maybe removing them isn't important for performance.

Separately it's a code-quality thing—if _some_ of our older functional components are using `useCallback` and some aren't, then it might be confusing as to the reason _why_ it's being used. That's also a reason to remove them I think, but it's worth pointing out it's not about performance, just about developer experience.

**Comment by aaemnnosttv** [comment]
@eugene-manuilov – I think we're on the same page so hear me out 😄  To clarify, using `useCallback` is itself a performance optimization which has its own cost in terms of computational resources, but also in terms of developer experience and overall complexity in a component. This hook in particular is somewhat easy to use in a way which introduces bugs due to a function's dependencies and those provided to the hook getting out of sync.

> I just want to make sure that we aren't going to spend a few dozen hours on reworking callback functions to make our scripts more performant on 0.00001%.

Strictly speaking about performance, any optimization resulting in a 0.00001% improvement wouldn't be worth doing. I don't think there is a substantial gain to be made by this change in that regard, but that's also kind of the point. We've already applied the optimization so that means we may have increased complexity without realizing actual gains 😄 
For me, this issue is more about simplifying our go-to approach in terms of component architecture by removing unnecessary optimizations. There is a bit of an assumption here that using `useCallback` or not will have _some_ impact on performance, so I agree that having something to measure is important to provide insight into what that impact is, if any.

Regarding the effort that it would take to change, I don't think it would take nearly that long (in many cases the wrapping hook could simply be removed) but that's not really the point here. If those hooks aren't providing a benefit that outweighs their cost, then we shouldn't use them because they add complexity.

This issue is also a starting point for us to discuss and start thinking about performance on the client, which hasn't really come up before as far as I know. What should we optimize for and which metrics are important? How should we be measuring them? What numbers should we be aiming for? How can we build this into our CI pipeline so that we have better visibility as to the implications of our changes? These are just a few questions that come to mind to start.

@kostyalmm @caseymorrisus thoughts?

**Comment by kostyalmm** [comment]
According to me, it is a code quality thing more than performance.
Maybe what we need here are comprehensive guidelines on when to use `useCallback`. We can pick up some community agreed standards and tweak them for our use case. 
Personally, what I feel odd is wrapping nearly every arrow function in `useCallback`.
A few wiki pages which can be referenced in PR comments will be great to have so the KT is easier for these kind of internal currently un-*written* rules.
I don't see the need for having a task to rewrite existing code. However, whenever someone touches a piece of code around `useCallback`s, they can either fix it(or comment in case of a PR).

**Comment by felixarntz** [comment]
@aaemnnosttv @tofumatt @eugene-manuilov Would be great to revive the conversation here. I don't think we should necessarily refactor all the `useCallback` usages to remove them if the benefits are minor; but it would be good to come to a conclusion here for moving forward (should we use `useCallback` or not, or if partially, when?).

**Comment by eugene-manuilov** [comment]
@felixarntz maybe let's arrange a call to discuss it and make a final decision?

**Comment by aaemnnosttv** [comment]
After discussing with @felixarntz, @tofumatt, and @eugene-manuilov, we decided it's more straightforward to use `useCallback` in all cases rather than deciding when we need to and when we don't which is arguably more complicated than solving the complexities related to hook dependencies and stale callbacks. This seems to be more consistent with the code in core/Gutenberg as well.

The plan is to leave existing code as-is and use `useCallback` for all callbacks going forward which aren't guaranteed to be stable. To support this and hopefully alleviate some of the burden in code review, we'll open a new issue to add additional tooling/configuration for ESLint to help with some of the heavy-lifting around ensuring dependency arrays are sufficiently complete.
